File: src/main/java/net/mcbrawls/blueprint/CompoundRegion.java
package net.mcbrawls.blueprint;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;
import kotlin.Unit;
import kotlin.jvm.functions.Function1;
import net.mcbrawls.blueprint.region.serialization.SerializableRegionTypes;
import net.mcbrawls.blueprint.serialization.SerializableRegion;
import net.minecraft.entity.Entity;
import net.minecraft.util.math.BlockPos;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.Set;

public class CompoundRegion extends SerializableRegion {
    public static final Codec<CompoundRegion> CODEC = RecordCodecBuilder.create(
            instance -> {
                return instance.group(
                        SerializableRegion.CODEC.listOf()
                                .fieldOf("regions")
                                .forGetter(CompoundRegion::getRegions)
                ).apply(instance, CompoundRegion::new);
            }
    );

    private final List<SerializableRegion> regions;

    public CompoundRegion(List<SerializableRegion> regions) {
        super(SerializableRegionTypes.INSTANCE.getCOMPOUND());
        this.regions = regions;
    }

    public List<SerializableRegion> getRegions() {
        return regions;
    }

    @NotNull
    @Override
    public Set<BlockPos> getPositions() {
        return null;
    }

    @Override
    public boolean contains(@NotNull Entity entity) {
        return false;
    }

    @Override
    public void forEachPosition(@NotNull Function1<? super BlockPos, Unit> action) {

    }
}

File: src/main/kotlin/net/mcbrawls/blueprint/region/CompoundRegion.kt
package net.mcbrawls.blueprint.region

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import dev.andante.codex.SetCodec.Companion.setOf
import dev.andante.codex.nullableFieldOf
import net.mcbrawls.blueprint.region.serialization.SerializableRegion
import net.mcbrawls.blueprint.structure.Blueprint
import net.minecraft.entity.Entity
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Vec3d

/**
 * A region defined of multiple other regions.
 * Useful for regions which cannot be defined through a single format.
 */
data class CompoundRegion(
    /**
     * The regions which compose this compound region.
     */
    val regions: Set<SerializableRegion>,

    /**
     * The global offset for this compound region.
     */
    val globalOffset: Vec3d? = null
) : Region {
    /**
     * A secondary vararg constructor.
     */
    constructor(
        vararg regions: SerializableRegion,
        globalOffset: Vec3d? = null
    ) : this(regions.toSet(), globalOffset)

    private fun getAbsoluteOffset(offset: Vec3d): Vec3d {
        return if (globalOffset != null) {
            offset.add(globalOffset)
        } else {
            offset
        }
    }

    override fun getBlockPositions(offset: Vec3d): Set<BlockPos> {
        val absoluteOffset = getAbsoluteOffset(offset)
        val mapped = regions.flatMap { region -> region.getBlockPositions(absoluteOffset) }
        return mapped.toSet()
    }

    override fun contains(entity: Entity, offset: Vec3d): Boolean {
        val absoluteOffset = getAbsoluteOffset(offset)
        return regions.any { region -> region.contains(entity, absoluteOffset) }
    }

    companion object {
        /**
         * The codec of a compound region.
         */
        val CODEC: Codec<CompoundRegion> = RecordCodecBuilder.create { instance ->
            instance.group(
                SerializableRegion.CODEC.setOf()
                    .fieldOf("regions")
                    .forGetter(CompoundRegion::regions),
                Vec3d.CODEC
                    .nullableFieldOf("offset")
                    .forGetter(CompoundRegion::globalOffset)
            ).apply(instance, ::CompoundRegion)
        }

        /**
         * Creates a compound region from the given regions of a blueprint.
         * @return a compound region
         */
        fun of(blueprint: Blueprint, vararg keys: String): CompoundRegion {
            val regions = keys.mapNotNull(blueprint.regions::get)
            return CompoundRegion(regions.toSet())
        }
    }
}
